llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -55,15 +56,91 @@ void InterBlockScheduling::markEpilogueBlocks() {
   }
 }
 
+std::map<MachineBasicBlock *, LivePhysRegs> calculateLiveOuts(MachineFunction *MF) {
+  std::queue<MachineBasicBlock *> WorkList;
+  std::set<MachineBasicBlock *> WorkListSet;
+  std::map<MachineBasicBlock *, LivePhysRegs> LiveIns;
+  std::map<MachineBasicBlock *, LivePhysRegs> LiveOuts;
+  const TargetRegisterInfo *TRI = MF->getSubtarget().getRegisterInfo();
+  for (MachineBasicBlock *MBB : post_order(MF)) {
```

Comment:
I get nervous when I see `post_order` on a DAG. There can be cycles, so I don't think we have the guarantee that all BB successors have been analyzed before we analyze BB itself.

I assume this is why you have a worklist and potentially run over the same BB multiple times. Maybe mention that using `post_order` is not about correctness, but rather optimizing the number of times we need to "re-run".

Reply:
To wit: entering them in forward order makes the algorithm weakly quadratic in the number of basic blocks. In forward flow you would normally just push the entry point and run the worklist. In backward flow this is problematic, because backward flow is not guaranteed to reach all blocks